compiler.py

```diff
@@ -0,0 +1,637 @@
+        #      different namespaces for different levels, but to allow that we'd also need to store old values for the
+        #      namespace in the stack too, which would be really taxing for slow mindustry processors!)
+
+        for subnode in node.body:
+            self.visit(subnode)
+
+        epilogue = len(self._ins)
+        # TODO better way to patch this (ins should be concrete types, not immutable strings)
+        #      Telos:  I've caught myself having this sort of thought many times over the years, but these strings are
+        #              so short that creating new ones with heavily optimized methods like format just is better/faster
+        #              than any sort of mutable data-type we'd be likely to come up with!
+        #              However, in this particular case, I think we should consider making each function name be an
+        #              mlog variable that points to the beginning of that function's body, so that "calling" the function
+        #              later is just a matter of "passing" variables, setting a return pointer, and then setting the
+        #              @counter to jump to the line number stored in that function's name.  But I think you'll still
+        #              be stuck going back to make one substitution later depending upon where exactly you end up
+        #              deciding to stick the function body.  Do you really need to iterate through *all* the lines
+        #              between prologue and epilogue?  You could make each Return jump straight back
+        #              to the caller, rather than detouring through the epilogue, and then the only remaining thing
+        #              you'd need to substitute for after the fact is the {} in the prologue.
+
+        for i in range(prologue, epilogue):
+            if '{epilogue}' in self._ins[i]:
+                self._ins[i] = self._ins[i].format(epilogue=epilogue)
+
+        # Add 1 to the return value to skip the jump that made the call.
+        self._ins.append(f'op add @counter {reg_ret} 1')
+
+        end = len(self._ins)
+        self._ins[prologue - 1] = self._ins[prologue - 1].format(end)
+        self._in_def = None
+
+    def visit_Return(self, node):
+        """This will be called for any instance of Return, typically from within a function definition."""
+        val = self.as_value(node.value)
+        self._ins.append(f'set {REG_RET} {val}')
+        self._ins.append('jump {epilogue} always')
+        # TODO  better to pass back the return value in a dummy variable, rather than requiring a stack
+        # TODO  maybe raise an error if not self._in_def  (can't return except when within a function definition)
+        #       what we have currently will end up leaving the {epilogue} in the compiled program
+        #       Or does AST itself enforce that Return must be within the scope of a Def?

```

Comment:
No, `ast` does not enforce this, though it's a bit weird because it is a `SyntaxError` to do it outside:

```python
>>> ast.dump(ast.parse('return'))
'Module(body=[Return()], type_ignores=[])'
>>> return
  File "<stdin>", line 1
SyntaxError: 'return' outside function
```